canonical-web-and-design/maas-ui
# feat(ui): build "Update certificate" flow in VM host config

## Done

- Renamed `KVMConfiguration` to `KVMSettings` and moved the configuration card to its own component in `/KVMSettings/KVMConfiguration` (this makes the diff appear larger than it is, sorry! Let me know if you prefer I split the file-shuffling to a separate PR).
- Built `UpdateCertificate` component in `AuthenticationCard` which handles generating/providing a new cert/key combo
- Extracted a couple of common components (`CertificateDownload` and `CertificateMetadata`) from the Add KVM and Edit KVM sections

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the KVM settings page of a KVM that has a certificate (e.g. `test` on bolla, or any newly created KVM)
- Click "Update certificate" and change the radio to provide a certificate
- Enter anything in the textareas and check that a `pod/update` action is dispatched with the correct payload. Note that errors are duplicated in the Authentication card and the KVM configuration card, to be addressed in #3055 
- Cancel the form and open it again by clicking "Update certificate"
- Leave the "Generate new certificate" radio selected and click "Next"
- Take note of the certificate data and click "Save"
- The form should close and the pod should have updated with the new certificate data 
- Go to the KVM settings page of a KVM that doesn't have a certificate (e.g. `another` on bolla)
- Check that the Authentication card shows the generate/provide cert radios automatically

## Fixes

Fixes canonical-web-and-design/app-squad#256

## Screenshot

![Peek 2021-09-16 22-44](https://user-images.githubusercontent.com/25733845/133614340-80e01508-0895-4704-a88c-b8954e6f11e7.gif)
